pkg/agent/billing/billing.go

```diff
@@ -146,31 +155,44 @@ func (s *metricsState) collect(conf *Config, store VMStoreForNode, metrics PromM
 	vmsOnThisNode := store.ListIndexed(func(i *VMNodeIndex) []*vmapi.VirtualMachine {
 		return i.List()
 	})
+	logger.Info(fmt.Sprintf("num_vms: %d", len(vmsOnThisNode)))
 	for _, vm := range vmsOnThisNode {
 		endpointID, isEndpoint := vm.Labels[EndpointLabel]
 		metricsBatch.inc(isEndpointFlag(isEndpoint), autoscalingEnabledFlag(api.HasAutoscalingEnabled(vm)), vm.Status.Phase)
-		if !isEndpoint {
-			// we're only reporting metrics for VMs with endpoint IDs, and this VM doesn't have one
-			continue
-		}
+		// if !isEndpoint {
+		// 	// we're only reporting metrics for VMs with endpoint IDs, and this VM doesn't have one
+		// 	continue
+		// }
 
 		if !vm.Status.Phase.IsAlive() || vm.Status.CPUs == nil {
+			logger.Info("Skipping gathering metrics for VM")
 			continue
 		}
 
+		byteCounts, err := vm.GetNetworkUsage()
+		if err != nil {
+			return err
+		}
+		logger.Info(fmt.Sprintf("Got byte counts %v", byteCounts))
 		key := metricsKey{
 			uid:        vm.UID,
 			endpointID: endpointID,
 		}
+
 		presentMetrics := vmMetricsInstant{
-			cpu: *vm.Status.CPUs,
+			cpu:          *vm.Status.CPUs,
+			ingressBytes: byteCounts.IngressBytes,
+			egressBytes:  byteCounts.EgressBytes,
 		}
+
 		if oldMetrics, ok := old[key]; ok {
 			// The VM was present from s.lastTime to now. Add a time slice to its metrics history.
 			timeSlice := metricsTimeSlice{
 				metrics: vmMetricsInstant{
 					// strategically under-bill by assigning the minimum to the entire time slice.
-					cpu: util.Min(oldMetrics.cpu, presentMetrics.cpu),
+					cpu:          util.Min(oldMetrics.cpu, presentMetrics.cpu),
+					ingressBytes: presentMetrics.ingressBytes,
+					egressBytes:  presentMetrics.egressBytes,
```

Comment:
Does `ingressBytes`/`egressBytes` get modified on each request? or should this subtract `oldMetrics` from `presentMetrics` to get the diff?

(And if it's the first, does that mean that if a request times out we won't see that usage?)

Reply:
Good catch, thank you! I thought I'd made the BPF zero out the counter every request, but I actually didn't. Actually it's better this way for the reason you said - if the request times out, we won't lose any count. I'll add the subtraction (though it does feel mildly silly to be subtracting and then re-adding inside of the history finalize).